MosaicResidentInformationApi/V1/Gateways/MosaicGateway.cs

```diff
@@ -15,35 +20,114 @@ public class MosaicGateway : IMosaicGateway
-                .GroupBy(address => address.Person, MapPersonAndAddressesToResidentInformation)
+        private List<long> PeopleIdsForAddressQuery(int cursor, int limit, string firstname, string lastname, string postcode, string address)
+        {
+            var firstNameSearchPattern = GetSearchPattern(firstname);
+            var lastNameSearchPattern = GetSearchPattern(lastname);
+            var addressSearchPattern = GetSearchPattern(address);
+            var postcodeSearchPattern = GetSearchPattern(postcode);
+            return _mosaicContext.Addresses
+                .Where(add =>
+                    string.IsNullOrEmpty(address) || EF.Functions.ILike(add.AddressLines.Replace(" ", ""), addressSearchPattern))
+                .Where(add =>
+                    string.IsNullOrEmpty(postcode) || EF.Functions.ILike(add.PostCode.Replace(" ", ""), postcodeSearchPattern))
+                .Where(add =>
+                    string.IsNullOrEmpty(firstname) || EF.Functions.ILike(add.Person.FirstName, firstNameSearchPattern))
+                .Where(add =>
+                    string.IsNullOrEmpty(lastname) || EF.Functions.ILike(add.Person.LastName, lastNameSearchPattern))
+                .Include(add => add.Person)
+                .Where(add => add.PersonId > cursor)
+                .GroupBy(add => add.PersonId)
+                .Where(p => p.Key != null)
+                .Take(limit)
+                .Select(p => (long) p.Key)
                 .ToList();
+        }
 
-            var peopleWithNoAddress = string.IsNullOrEmpty(postcode) && string.IsNullOrEmpty(address)
-                ? QueryPeopleWithNoAddressByName(firstname, lastname, addressesFilteredByPostcode, cursor)
-                : new List<ResidentInformation>();
+        private static List<PhoneNumber> MapPhoneNumbers(IEnumerable<TelephoneNumber> z)
+        {
+            var phoneNumberList = z.Where(tn => tn.Number != null && tn.Type != null).Select(tn => new PhoneNumber
+            {
+                Number = tn.Number,
+                Type = Enum.Parse<PhoneType>(tn.Type)
+            }).Distinct().ToList();
 
-            var allPeople = peopleWithAddresses.Concat(peopleWithNoAddress);
+            return !phoneNumberList.Any() ? null : phoneNumberList;
+        }
 
-            return allPeople.Select(AttachPhoneNumberToPerson).OrderBy(a => a.MosaicId).Take(limit).ToList();
+        private static List<DomainAddress> MapAddress(IEnumerable<Address> q)
+        {
+            var addressList = q.Where(a => a.AddressLines != null && a.PostCode != null).Select(a => new DomainAddress
+            {
+                AddressLine1 = a.AddressLines,
+                PostCode = a.PostCode
+            }).Distinct().ToList();
+
+            return !addressList.Any() ? null : addressList;
         }
 
-        public ResidentInformation GetEntityById(int id)
+        public ResidentInformation GetEntityById(long id)
```

Comment:
Can we move the public method to be above the private ones please

Reply:
Sure, will do that.

Reply:
Done in d454b7a